### ISSUE TITLE
tf.layers.conv3d with "channels_first" does not accept dynamic shapes

### ISSUE THREAD

**Issue by ORippler** [tensorflow/tensorflow#20527]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
binary
- **TensorFlow version (use command below)**:
v1.8.0-0-g93bc2e2072
- **Python version**: 
3.6
- **Bazel version (if compiling from source)**:
N/A
- **GCC/Compiler version (if compiling from source)**:
N/A
- **CUDA/cuDNN version**:
cuDNN 7.0 CUDA 9.1
- **GPU model and memory**:
Geforce 1080 Ti 12Gb
- **Exact command to reproduce**:
See given code snippets

### Describe the problem
Calling tf.layers.conv3d with "channels_first" on a tf.placeholder with dynamic shapes produces an error, whereas it does not for "channels_last" (see source code below). Cause is usage of reshape operation [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/convolutional.py#L203-L207) as outlined in a similar [issue](https://github.com/tensorflow/tensorflow/issues/15655). Unfortunately, no general solution supporting dynamic shapes was found in aforementioned issue, and shape inference does not allow for >1 dynamic shape.

### Source code / logs
this produces error:
```
import tensorflow as tf
x = tf.placeholder(dtype=tf.float32, shape=[None, 1, None, None, None])
y = tf.layers.conv3d(x, 32, 9, data_format='channels_first')
```

this works fine:
```
import tensorflow as tf
x = tf.placeholder(dtype=tf.float32, shape=[None, None, None, None, 1])
y = tf.layers.conv3d(x, 32, 9, data_format='channels_last')
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Exact command to reproduce

**Comment by Schnabbel** [comment]
@tensorflowbutler done

**Comment by facaiy** [comment]
If `bias_add` supports 5-d tensor for NCHW, we could remove the reshape operator which seems error-prone with dynamic shape.

**Comment by Schnabbel** [comment]
According to [documentation,](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/ops/nn_ops.py#L1485-L1488) nn.bias_add supports N-d tensors due to broadcasting.

However, the underlying [C implementation/documentation](https://github.com/tensorflow/tensorflow/blob/86f5ab7474825da756838b34e1b4eac93f5fc68a/tensorflow/core/ops/nn_ops.cc#L430-L455)  states explicitly that for NCHW, bias would be added to third-to-the-last dimension, which forces the reshape operations.

I would therefore suggest a change here, such that bias is added to first dimension instead of third-to-the-last dimension, as channels will always be first dim in NCHW format. This would make reshaping operations obsolete.

**Comment by facaiy** [comment]
Source code is [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/bias_op.cc). I think you can improve it if you'd like :-)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by yongtang** [comment]
Added a PR #22127 for 3D/4D/5D support of bias_add with NCHW data format.

**Comment by facaiy** [comment]
Create a PR #23004 to fix the issue.

**Comment by yongtang** [comment]
Close the issue as PRs have been merged.
